### ISSUE TITLE
Slashing information validation

### ISSUE THREAD

**Issue by linked0** [bosagora/agora#1444]
This issue is related to #1076 which is about implementing the slashing protocol. We need to split some tasks related to block header validation from issue #1076 because the tasks have many things to do and could consume some time.
 
We've been adding information about slashed validators into a block. The newly added information is as follows.

```
/// Hash of random seed of the preimages for this this height
Hash random_seed;

/// List of indices to the validator UTXO set which have not revealed the preimage
uint[] missing_validators;
```

The `random_seed` must not null in any case, which is generated from the preimages of current validators, and it must be the same as the local `random_seed` from a node itself. It's the things for the validation routine to check. In the current code of Agora, there are 2 cases the information is not valid as follows besides the case where a misbehaving validator manipulates the information.

1. In the catch-up process, there are no pre-images for generating the local random seed needed for the comparison. So we need a process to get pre-images from other nodes and add the information into the `ValidatorSet`.
2. In many unit tests or network tests, there are some cases that do not consider maintaining preimages needed for validation check and generating the `random_seed`.

As a goal of this issue, we have to add the validation routine about slashing information of the block header and make the code pass the above situations and tests.

**Comment by linked0** [comment]
This is what I want to talk about in a **whiteboarding session**. Please let me know your opinions. Thanks.

There's one thing we should think about the catch-up process and validation of slashing information. When the catch-up process is running in a node, the node needs to know the validator's pre-images because it needs to check the validity of the slashing information in each block. So it calls the `getPreimages` API endpoint and gets preimages from the network. But, what if there's no node from which we can get pre-images? We can imagine that there are no valid nodes like they only have the block data and no normal database for validators. The situation I described here is very critical and has a very low possibility, which means AGORA network doesn't work anymore. But it might happen often in the test network. 

Don't we need to store the last pre-images of un-enrolled validators into the ledger or any other methods, or it's just a simple issue that can be handled with some `if` condition?

**Comment by AndrejMitrovic** [comment]
> But, what if there's no node from which we can get pre-images?

Either:
- keep trying from all the nodes in the node registry
- find another node registry
- the network is really shut down

I don't understand how this is a slashing problem. It's the same problem as if nodes refuse to send you blocks, isn't it?

**Comment by linked0** [comment]
> I don't understand how this is a slashing problem. It's the same problem as if nodes refuse to send you blocks, isn't it?

I agree with your point. But, my concern was that we had needed only valid blocks before adding validation checks for slashing information but we need valid blocks and valid data on the database after adding that. Even if this is not a slashing problem, I think we need to consider the case when only the block files remain but there is no database in the TestNet, don't we? I will think about it more.

**Comment by AndrejMitrovic** [comment]
Well the big benefit of https://github.com/bpfkorea/agora/pull/1421 is that we only need to store the very last preimage of any UTXO in the database. So the total amount of data needed to be stored is very low.

**Comment by linked0** [comment]
@AndrejMitrovic Thanks for all your opinions. We agree that the situation where there's no node from which we can get pre-images is not a slashing problem as describe in [this comment](https://github.com/bpfkorea/agora/issues/1444#issuecomment-763265894), so this issue can be solved with the PR #1468 merged. And, I think we don't need to have a white-boarding session about this issue. 

@Geod24 If there is something more to do for PR #1468 to be approved or you have other opinions, please let me know.
@bpalaggi We can meet the deadline about the slashing protocol if this issue is closed, although there are small issues about the slashing like issue #1487
